File: solutions/r/bob/bob.R
# Bob returns a string that only ever answers one of five things.
#
# Bob is a lackadaisical teenager. He likes to think that he's very cool.
# And he definitely doesn't get excited about things. That wouldn't be cool.
bob <- function(input) {
    reply <- list(
        default = "Whatever.",
        question = "Sure.",
        yelling = "Whoa, chill out!",
        yellingQuestion = "Calm down, I know what I'm doing!",
        silence = "Fine. Be that way!"
    )

    input <- trimws(input)
    if (input == "") {
        return(reply$silence)
    }

    is_yelling <- input == toupper(input) && input != tolower(input)
    is_question <- substr(input, nchar(input), nchar(input)) == "?"

    if (is_yelling && is_question) {
        reply$yellingQuestion
    } else if (is_yelling) {
        reply$yelling
    } else if (is_question) {
        reply$question
    } else {
        reply$default
    }
}

File: solutions/javascript/bob/bob.js
// @ts-check

// Bob is a lackadaisical teenager. He likes to think that he's very cool.
// And he definitely doesn't get excited about things. That wouldn't be cool.
const reply = {
  default: "Whatever.",
  question: "Sure.",
  yelling: "Whoa, chill out!",
  escalate: "Calm down, I know what I'm doing!",
  silence: "Fine. Be that way!",
};

/**
 * Hey returns a string that only ever answers one of five things.
 * @param {string} message
 * @return {string} response
 */
export const hey = (message) => {
  const isSilence = !/\S/.test(message);
  if (isSilence) return reply.silence;

  const isYelling = /[A-Z]/.test(message) && !/[a-z]/.test(message);
  const isQuestion = /\?\s*$/.test(message);

  return isYelling && isQuestion
    ? reply.escalate
    : isYelling
    ? reply.yelling
    : isQuestion
    ? reply.question
    : reply.default;
};
